examples/async_api_check.py

```diff
@@ -0,0 +1,177 @@
+
+
+def handle_done_pending_tasks(done, pending):
+    for done_task in done:
+        try:
+            done_task.result()
+        except asyncio.CancelledError:
+            logger.warning(f"Cancelled: {done_task}")
+
+    for pending_task in pending:
+        try:
+            pending_task.cancel()
+        except asyncio.CancelledError:
+            # cancelling is the intended behaviour
+            pass
+
+
+class DeviceDiscoverer:
+    def __init__(self, search_timeout):
+        self.selected_device_info = None
+        self.search_timeout = search_timeout
+
+    async def get_user_selected_device(self):
+        async with Network() as network:
+            print("Looking for devices in your network...\n\t", end="")
+            await network.wait_for_new_device(timeout_seconds=self.search_timeout)
+
+            while self.selected_device_info is None:
+                print("\n======================================")
+                print("Please select a Pupil Invisible device by index:")
+                for device_index, device_name in enumerate(network.devices):
+                    print(f"\t{device_index}\t{device_name}")
+
+                print("To reload the list, hit enter.")
+                user_input = await input_async()
+                self.selected_device_info = evaluate_user_input(
+                    user_input, network.devices
+                )
+
+
+class DataReceiver:
+    def __init__(self, device_info):
+        self.device_info = device_info
+        self.notifier = None
+        self.status = None
+        self.gaze_sensor_url = None
+        # self.world_sensor = None
+
+    async def update_status(self):
+        async with Device.from_discovered_device(self.device_info) as device:
+            self.status = await device.get_status()
+            self.gaze_sensor_url = self.status.direct_gaze_sensor().url
+            # self.world_sensor = self.status.direct_world_sensor()
+
+    async def on_update(self, component):
+        if isinstance(component, Sensor):
+            if component.sensor == 'gaze' and component.conn_type == 'DIRECT':
+                await self.update_status()
+
+    async def make_status_update_notifier(self):
+        async with Device.from_discovered_device(self.device_info) as device:
+            self.notifier = StatusUpdateNotifier(device, callbacks=[self.on_update])
+            await self.notifier.receive_updates_start()
+
+    async def cleanup(self):
+        await self.notifier.receive_updates_stop()
+
+
+class Adapter:
+    def __init__(self, selected_device):
+        self.receiver = DataReceiver(selected_device)
+        self.publisher = pi_gaze_relay.PupilInvisibleGazeRelay()
+        self.gaze_sample_queue = asyncio.Queue()
+        self.publishing_task = None
+        self.receiving_task = None
+
+    async def receive_gaze_sample(self):
+        while True:
+            if self.receiver.gaze_sensor_url:
+                async for gaze in receive_gaze_data(
```

Comment:
I wonder if there should be some kind of "receive gaze as soon as possible" API, directly built into the Realtime API module. That would simplify this use case and I can foresee others needing it too.